File: ios/native/App/App/DriveLink.swift
import AVFoundation
import Foundation
import WebKit

/// The single channel between the CarPlay templates on the car screen and the
/// `/drive` page running in the app's WebView on the phone.
///
/// CarPlay's voice-based conversational category does not hand an app a drawing
/// surface: the car screen is Apple's templates, and nothing else. So the
/// conversation, the agent, its voice and its 3D face all stay in the WebView,
/// and this object is the wire between them. The page posts its state out
/// (`window.webkit.messageHandlers.threeWsDrive`), and template presses go back
/// in as `window.threeWsDrive.command({ type })`. Both halves of that protocol
/// live in `src/drive/bridge.js`.
///
/// It also owns the audio session, because CarPlay requires a conversational
/// app to hold one open only while voice is actually in use.
final class DriveLink: NSObject {

    static let shared = DriveLink()

    enum State: String {
        case idle, listening, transcribing, thinking, speaking

        /// Is a turn in flight? The car screen shows the voice template for
        /// exactly this window and returns to the list outside it.
        var isActive: Bool { self != .idle }
    }

    struct Action {
        let id: String
        let label: String
        let enabled: Bool
    }

    /// Name of the WebKit message handler. Must match `src/drive/bridge.js`.
    private static let channel = "threeWsDrive"
    private static let protocolVersion = 1

    private(set) var state: State = .idle
    private(set) var agentName: String?
    private(set) var lastSpoken: String?
    private(set) var lastHeard: String?
    private(set) var lastError: String?
    private(set) var actions: [Action] = []

    /// Called on the main queue whenever anything above changes.
    var onChange: (() -> Void)?

    private weak var webView: WKWebView?
    private var audioSessionActive = false

    // MARK: - Attachment

    /// Register on the app's WebView. Safe to call more than once.
    func attach(to webView: WKWebView) {
        if self.webView === webView { return }
        detach()
        self.webView = webView
        let controller = webView.configuration.userContentController
        controller.removeScriptMessageHandler(forName: Self.channel)
        controller.add(self, name: Self.channel)
    }

    func detach() {
        webView?.configuration.userContentController.removeScriptMessageHandler(forName: Self.channel)
        webView = nil
    }

    // MARK: - Outbound: car screen to page

    /// Send a command the page understands. See `onNativeCommand` in
    /// `src/drive/index.js` for the accepted types.
    func send(_ type: String, value: String? = nil) {
        guard let webView else { return }
        var payload: [String: Any] = ["type": type]
        if let value { payload["value"] = value }
        guard
            let data = try? JSONSerialization.data(withJSONObject: payload),
            let json = String(data: data, encoding: .utf8)
        else { return }
        // Optional chaining on the page side: a command that arrives before the
        // page has booted is dropped rather than throwing into the console.
        webView.evaluateJavaScript("window.threeWsDrive?.command(\(json));", completionHandler: nil)
    }

    /// Point the phone's WebView at the car surface. Called when the CarPlay
    /// scene connects, which only happens when the driver launches three.ws
    /// from the CarPlay home screen: navigating then is what they asked for.
    /// The previous page is remembered so disconnecting puts them back.
    func openDriveSurface() {
        guard let webView else { return }
        let current = webView.url
        if current?.path.hasPrefix("/drive") != true {
            returnURL = current
        }
        var components = URLComponents()
        components.scheme = "https"
        components.host = current?.host ?? "three.ws"
        components.path = "/drive"
        components.queryItems = [URLQueryItem(name: "surface", value: "carplay")]
        guard let url = components.url else { return }
        webView.load(URLRequest(url: url))
    }

    /// Put the phone back where the driver left it when CarPlay disconnects.
    func closeDriveSurface() {
        guard let webView, let returnURL else { return }
        self.returnURL = nil
        webView.load(URLRequest(url: returnURL))
    }

    private var returnURL: URL?

    // MARK: - Audio session

    /// CarPlay requires a conversational app to hold an audio session only
    /// while voice is actually in use, so this is driven by the page's own
    /// state rather than held open for the whole trip.
    private func syncAudioSession() {
        let shouldBeActive = state.isActive
        guard shouldBeActive != audioSessionActive else { return }
        let session = AVAudioSession.sharedInstance()
        do {
            if shouldBeActive {
                try session.setCategory(
                    .playAndRecord,
                    mode: .voiceChat,
                    options: [.allowBluetooth, .allowBluetoothA2DP, .duckOthers]
                )
                try session.setActive(true)
            } else {
                try session.setActive(false, options: .notifyOthersOnDeactivation)
            }
            audioSessionActive = shouldBeActive
        } catch {
            // A refused session must not take the conversation down: the page
            // surfaces its own microphone error, and the next turn retries.
            NSLog("[drive] audio session %@ failed: %@", shouldBeActive ? "activate" : "deactivate", error.localizedDescription)
        }
    }

    // MARK: - Reset

    func reset() {
        state = .idle
        agentName = nil
        lastSpoken = nil
        lastHeard = nil
        lastError = nil
        actions = []
        syncAudioSession()
    }
}

// MARK: - Inbound: page to car screen

extension DriveLink: WKScriptMessageHandler {

    func userContentController(_ controller: WKUserContentController, didReceive message: WKScriptMessage) {
        guard message.name == Self.channel,
              let body = message.body as? [String: Any],
              (body["v"] as? Int) == Self.protocolVersion,
              let type = body["type"] as? String
        else { return }

        switch type {
        case "ready":
            let agent = body["agent"] as? [String: Any]
            agentName = agent?["name"] as? String
            lastError = nil
        case "state":
            state = State(rawValue: body["state"] as? String ?? "") ?? .idle
            syncAudioSession()
        case "heard":
            lastHeard = body["text"] as? String
        case "said":
            lastSpoken = body["text"] as? String
            lastError = nil
        case "error":
            lastError = body["message"] as? String
        case "actions":
            let raw = body["actions"] as? [[String: Any]] ?? []
            actions = raw.compactMap { entry in
                guard let id = entry["id"] as? String, let label = entry["label"] as? String else { return nil }
                return Action(id: id, label: label, enabled: (entry["enabled"] as? Bool) ?? true)
            }
        default:
            return
        }

        onChange?()
    }
}

File: solana-mobile/android-overlay/app/src/main/java/ws/three/app/drive/DriveLink.java
package ws.three.app.drive;

import android.os.Handler;
import android.os.Looper;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * The single channel between the Android Auto templates on the car screen and
 * the /drive page running in {@link DriveWebService}'s web view.
 *
 * Android Auto's template host renders the car screen; an app never draws on it
 * outside the navigation category, and three.ws is not a navigation app. So the
 * conversation, the agent, its voice and its memory all stay in the page, and
 * this object is the wire between them. The page posts its state out through the
 * {@code ThreeWsDriveNative} JavaScript interface, and template presses go back
 * in as {@code window.threeWsDrive.command(...)}. Both halves of that protocol
 * live in src/drive/bridge.js, and the iOS half is DriveLink.swift.
 *
 * Every mutation lands on the main looper, because the page calls in from the
 * WebView's JavaScript thread and the car screen may only be invalidated from
 * the main thread.
 */
public final class DriveLink {

    /** Protocol version. Must match PROTOCOL in src/drive/bridge.js. */
    static final int PROTOCOL = 1;

    public static final String STATE_IDLE = "idle";

    /** One control mirrored onto the car screen. */
    public static final class Action {
        public final String id;
        public final String label;
        public final boolean enabled;

        Action(String id, String label, boolean enabled) {
            this.id = id;
            this.label = label;
            this.enabled = enabled;
        }
    }

    /** Implemented by the car screen so it can repaint when anything changes. */
    public interface Listener {
        void onDriveChanged();
    }

    /** Implemented by the web service: delivers a command into the page. */
    public interface CommandSink {
        void sendCommand(String json);
    }

    private static final DriveLink INSTANCE = new DriveLink();

    public static DriveLink get() {
        return INSTANCE;
    }

    private final Handler main = new Handler(Looper.getMainLooper());

    private String state = STATE_IDLE;
    private String agentName;
    private String lastSpoken;
    private String lastHeard;
    private String lastError;
    private List<Action> actions = Collections.emptyList();

    private Listener listener;
    private CommandSink sink;

    private DriveLink() {
    }

    // -- reads (main thread) --------------------------------------------------

    public String state() {
        return state;
    }

    /** Is a turn in flight? The car screen shows progress for exactly this window. */
    public boolean isActive() {
        return !STATE_IDLE.equals(state);
    }

    public String agentName() {
        return agentName;
    }

    public String lastSpoken() {
        return lastSpoken;
    }

    public String lastHeard() {
        return lastHeard;
    }

    public String lastError() {
        return lastError;
    }

    public List<Action> actions() {
        return actions;
    }

    // -- wiring ---------------------------------------------------------------

    public void setListener(Listener value) {
        listener = value;
    }

    public void setCommandSink(CommandSink value) {
        sink = value;
    }

    /**
     * Send a command the page understands. See onNativeCommand in
     * src/drive/index.js for the accepted types. Dropped, deliberately, when no
     * page is attached yet: a press before the loop is up is not an error.
     */
    public void send(String type) {
        CommandSink target = sink;
        if (target == null) return;
        try {
            JSONObject payload = new JSONObject();
            payload.put("type", type);
            target.sendCommand(payload.toString());
        } catch (Exception ignored) {
            // A malformed command must never take the drive down.
        }
    }

    public void reset() {
        main.post(new Runnable() {
            @Override
            public void run() {
                state = STATE_IDLE;
                agentName = null;
                lastSpoken = null;
                lastHeard = null;
                lastError = null;
                actions = Collections.emptyList();
                notifyChanged();
            }
        });
    }

    // -- inbound: page to car screen ------------------------------------------

    /**
     * Called from the JavaScript interface thread with one message from the
     * page. Anything unparseable is dropped rather than thrown: the page is a
     * remote document and this is a trust boundary.
     */
    void accept(String json) {
        final JSONObject body;
        try {
            body = new JSONObject(json);
        } catch (Exception e) {
            return;
        }
        if (body.optInt("v", -1) != PROTOCOL) return;
        final String type = body.optString("type", "");
        if (type.isEmpty()) return;

        main.post(new Runnable() {
            @Override
            public void run() {
                apply(type, body);
                notifyChanged();
            }
        });
    }

    private void apply(String type, JSONObject body) {
        switch (type) {
            case "ready": {
                JSONObject agent = body.optJSONObject("agent");
                agentName = agent == null ? null : emptyToNull(agent.optString("name", ""));
                lastError = null;
                break;
            }
            case "state":
                state = body.optString("state", STATE_IDLE);
                break;
            case "heard":
                lastHeard = emptyToNull(body.optString("text", ""));
                break;
            case "said":
                lastSpoken = emptyToNull(body.optString("text", ""));
                lastError = null;
                break;
            case "error":
                lastError = emptyToNull(body.optString("message", ""));
                break;
            case "actions":
                actions = readActions(body.optJSONArray("actions"));
                break;
            default:
                break;
        }
    }

    private static List<Action> readActions(JSONArray raw) {
        if (raw == null) return Collections.emptyList();
        List<Action> parsed = new ArrayList<>(raw.length());
        for (int i = 0; i < raw.length(); i++) {
            JSONObject entry = raw.optJSONObject(i);
            if (entry == null) continue;
            String id = entry.optString("id", "");
            String label = entry.optString("label", "");
            if (id.isEmpty() || label.isEmpty()) continue;
            parsed.add(new Action(id, label, entry.optBoolean("enabled", true)));
        }
        return Collections.unmodifiableList(parsed);
    }

    private void notifyChanged() {
        Listener target = listener;
        if (target != null) target.onDriveChanged();
    }

    private static String emptyToNull(String value) {
        return value == null || value.isEmpty() ? null : value;
    }
}
